### DISCUSSION TITLE
Throttled by AAD

### DISCUSSION THREAD

**Discussion by SuleymanA** [pnp/pnpcore#1024] [Q&A] [answered]
Hi,

I am using PnP Core in an Azure Function v4 (isolated). Mainly connecting to SP to perform actions. Authentication is done using "UsernamePasswordAuthenticationProvider". We received the following error while creating a pnp context like "using (var productContext = this._pnpContextFactory.Create(new Uri(productSiteUrl), authProvider))". Error message was: "Your app has been throttled by AAD due to too many requests. To avoid this, cache your tokens see https://aka.ms/msal-net-throttling". We are using the PnP multi-tenant app. 

I am trying to understand why this is happening and how to prevent it. Questions:
- Can this be prevented by using different app then the PnP multi-tenant app? 
- Is the really related to the app or the service account? 
- Should I do something with token caching? We are connecting to different sites so context is created on each call of the Azure function. 
- Is using the PnP multi-tenant app risky in terms of throttling because it will also be used by other processes like with using PnP Powershell?

Any help is much appreciated.

**Comment by jansenbe** [comment, marked as answer]
@SuleymanA : if there are other apps all using the PnP management shell app then this will add up and might get you throttled sooner. For use in functions it's recommended to implement a token cache as part of MSAL. See https://learn.microsoft.com/en-us/azure/active-directory/develop/msal-net-token-cache-serialization?tabs=aspnet#simple-token-cache-serialization-msal-only to learn more.

**Reply by SuleymanA** [reply]
@jansenbe thx for your reply. Do you know of an example project for using token cache in combination with PnP Core? All example I see are for ASP.Net Core web apps. I was lokking into the "UsernamePasswordAuthenticationProvider" and looks like it already using token cache that is available in "Microsoft.Identity.Client":
`                // Try to get the token from the tokens cache
                tokenResult = await publicClientApplication.AcquireTokenSilent(scopes, account.FirstOrDefault())
                    .ExecuteAsync().ConfigureAwait(false);`
Am I maybe missing some configuration to make this token cache work?

**Reply by jansenbe** [reply]
@SuleymanA : please check this sample: https://pnp.github.io/pnpcore/demos/Demo.PersistentTokenCache/README.html.

**Reply by SuleymanA** [reply]
Thx. I actually don't need a persistent token cache. The in memory cache of MSAL should be enough. But I think I managed to find a good solution...

I was instantiating the authentication provider in the function execution because the password changes everyday. I now moved the instantiation of the authentication provider to the startup of the Azure Function and pass the authentication provider on to the functions as a singleton ( to make sure that the same MSAL client is used). The password is refreshed in function execution to make sure that the right password is used. It looks like the number of singins is reduced. Also switched to using our own service principal for authentication.

Do you have any feedback on this setup (authentication provider instantiation in startup, pass it on as singleton and use in function execition)?

**Reply by jansenbe** [reply]
@SuleymanA : this is the right approach!
